src/aof.c

```diff
@@ -548,12 +557,101 @@ int persistAofManifest(aofManifest *am) {
     }
 
     sds amstr = getAofManifestAsString(am);
-    int ret = writeAofManifestFile(amstr);
+    int ret = writeAofManifestFile(server.aof_filename, amstr);
     sdsfree(amstr);
     if (ret == C_OK) am->dirty = 0;
     return ret;
 }
 
+/* Called in `loadAppendOnlyFiles` when we upgrade from a old version redis.
+ * 
+ * 1) Check if the UPGRADE_TEMP_DIR_NAME directory already exists.
+ *    1a) If it exists, then check if 'server.aof_filename' file exists in this 
+ *        directory, If it exists, then continue to execute rename and return.
+ *    1b) If it exists and the 'server.aof_filename' file not exists in this 
+ *        directory, then we directly delete UPGRADE_TEMP_DIR_NAME directory.
+ * 2) Create UPGRADE_TEMP_DIR_NAME directory.
+ * 3) Use 'server.aof_filename' to construct a BASE type aofInfo and add it to 
+ *    aofManifest, then persist the manifest file to UPGRADE_TEMP_DIR_NAME directory.
+ * 4) Move the old AOF file (server.aof_filename) to UPGRADE_TEMP_DIR_NAME directory.
+ * 5) rename UPGRADE_TEMP_DIR_NAME to server.aof_filename.
+ */
+int aofUpgradePrepare(aofManifest *am) {
+    sds aof_filepath = makePath(UPGRADE_TEMP_DIR_NAME, server.aof_filename);
+    /* If the UPGRADE_TEMP_DIR_NAME directory already exists, then we are in an 
+     * interrupted upgrade state. */
+    if (dirExists(UPGRADE_TEMP_DIR_NAME)) {
+        sds aof_filepath = makePath(UPGRADE_TEMP_DIR_NAME, server.aof_filename);
+        /* If the old AOF file has been successfully copied, we can directly
+         * perform the rename operation */
+        if (fileExist(aof_filepath)) {
+            sdsfree(aof_filepath);
+            goto rename;
+        } else {
+            /* Otherwise, we directly delete UPGRADE_TEMP_DIR_NAME dir and re-start
+             * the upgrade process. */
+            if (dirRemove(UPGRADE_TEMP_DIR_NAME) != 0) {
+                serverLog(LL_NOTICE, "Failed to remove append-only dir %s",
+                    UPGRADE_TEMP_DIR_NAME);
+                goto cleanup;
```

Comment:
does that cause a restart of the upgrade? looks like we're exiting with an error?
maybe i'm missing something (looking at the diff of this commit in GH rather than the full source tree)

Reply:
ohh, looking at this again, i realize the error is not related to the comment above it (the upgrade restarts only if we successfully remove the dir, but if we can't then we fail and exit.
i suppose this is not expected to ever happen.